### ISSUE TITLE
md-dialog: delayed focus on opening dialog

### ISSUE THREAD

**Issue by clshortfuse** [angular/material#9273]
**Actual Behavior**:
 -  `What is the issue? *` dialogs have a long animation delay when opening and the default button is not focused until animation closed.
 -  `What is the expected behavior?` default button should have focus during it's animation state

**CodePen** (or steps to reproduce the issue): *
 -  `CodePen Demo which shows your issue:` https://material.angularjs.org/latest/demo/dialog
 -  `Details:` Click Alert Dialog and press (Enter) not too long after. If you don't wait until the OK button turn gray (have focus), it'll click again.

**Angular Versions**: *
 -  `Angular Version:` 1.5.5
 -  `Angular Material Version:` 1.1.0-rc5 (master)

**Additional Information**:
 -  `Browser Type: *` Chrome
 -  `Browser Version: *` V52
 -  `OS: *` ChromeOS
 -  `Stack Traces:`

----

This is an accessibility issue where the dialog is fully present to the user, but the user must wait for a  before committing an action. This may sound trivial for dialog buttons where the user rarely ever sees, but for repeated tasks (especially those that use prompts based on keyevents), it becomes rather confusing.

For comparison, check out Chrome's `window.alert()` system. The OK button gets focus before the animation is complete, and you can press (Enter) before animation is done. Custom dialogs probably need the ability to disable this feature, but our default ones should work like this. 

At the very, whatever had focused should be trapped to ensure that if you hit Enter early, you're not sending events to the previous element. Holding the events and waiting a digest cycle after animation ends is also another option, though I honestly rather see focus without waiting for the animation.

https://www.w3.org/TR/wai-aria-practices/#trap_focus

**Comment by clshortfuse** [comment]
From: https://www.w3.org/TR/wai-aria-practices/#make_modal

>Before the dialog is shown, create and display the dialog underlay. Connect an onkeypress event handler to the DOM document.documentElement. This will catch all keystrokes on the document and allow trapping keyboard focus within the dialog. Size and position the dialog box in the viewport above the underlay, make it visible and set focus to the first tab focusable item in the dialog box.

The key here is to hook keypress events before making it visible.

**Comment by DevVersion** [comment]
So I tried to work on that the last days and it seems to require a lot of work.

> To get it working we need to rework the `md-autofocus` directive, because if we want to auto-focus at the time the animation starts the attribute expression needs to be evaluated before the next digest runs.

> Also the dialog template **might** be not ready at this time and makes the autofocus target invalid and means that we would need to wait for the digest to finish.

I consider this as a very unelegant approach and I don't want to make that change to the dialog yet.

> I will keep track of it and will take another look if the dialog uses the panel.
